core/src/avm1/globals/rectangle.rs

```diff
@@ -0,0 +1,1156 @@
+use crate::avm1::return_value::ReturnValue;
+use crate::avm1::{Avm1, Error, Object, ScriptObject, TObject, Value};
+use crate::context::UpdateContext;
+use enumset::EnumSet;
+use gc_arena::MutationContext;
+use std::f64::NAN;
+
+fn constructor<'gc>(
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    if args.is_empty() {
+        this.set("x", 0.into(), avm, context)?;
+        this.set("y", 0.into(), avm, context)?;
+        this.set("width", 0.into(), avm, context)?;
+        this.set("height", 0.into(), avm, context)?;
+    } else {
+        this.set(
+            "x",
+            args.get(0).unwrap_or(&Value::Undefined).to_owned(),
+            avm,
+            context,
+        )?;
+        this.set(
+            "y",
+            args.get(1).unwrap_or(&Value::Undefined).to_owned(),
+            avm,
+            context,
+        )?;
+        this.set(
+            "width",
+            args.get(2).unwrap_or(&Value::Undefined).to_owned(),
+            avm,
+            context,
+        )?;
+        this.set(
+            "height",
+            args.get(3).unwrap_or(&Value::Undefined).to_owned(),
+            avm,
+            context,
+        )?;
+    }
+
+    Ok(Value::Undefined.into())
+}
+
+fn to_string<'gc>(
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    let x = this
+        .get("x", avm, context)?
+        .resolve(avm, context)?
+        .coerce_to_string(avm, context)?;
+    let y = this
+        .get("y", avm, context)?
+        .resolve(avm, context)?
+        .coerce_to_string(avm, context)?;
+    let width = this
+        .get("width", avm, context)?
+        .resolve(avm, context)?
+        .coerce_to_string(avm, context)?;
+    let height = this
+        .get("height", avm, context)?
+        .resolve(avm, context)?
+        .coerce_to_string(avm, context)?;
+
+    Ok(format!("(x={}, y={}, w={}, h={})", x, y, width, height).into())
+}
+
+pub fn create_rectangle_object<'gc>(
+    gc_context: MutationContext<'gc, '_>,
+    rectangle_proto: Option<Object<'gc>>,
+    fn_proto: Option<Object<'gc>>,
+) -> Object<'gc> {
+    FunctionObject::function(
+        gc_context,
+        Executable::Native(constructor),
+        fn_proto,
+        rectangle_proto,
+    )
+}
+
+fn is_empty<'gc>(
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    let width = this
+        .get("width", avm, context)?
+        .resolve(avm, context)?
+        .as_number(avm, context)?;
+    let height = this
+        .get("height", avm, context)?
+        .resolve(avm, context)?
+        .as_number(avm, context)?;
+    Ok((width == 0.0 || height == 0.0 || width.is_nan() || height.is_nan()).into())
```

Comment:
Seems like this should be <= 0.0, `new Rectangle(1, 1, -1, -1).isEmpty()` should return true.

Reply:
Good catch, adding to the tests.